File: tools/pkf-decode-py/load.py
"""Port of the succinct-loading slice of proskomma-core's index.js
(loadSuccinctDocSet/newDocumentFromSuccinct) + model/doc_set.js
(fromSuccinct) — only the fields decode.mjs's actual GraphQL query shape
reads (see unsuccinctify.py's docstring): per block, only `bs`/`bg`/`c`;
`nt`/`os`/`is` are decoded by proskomma-core too but never queried by
PerfRenderFromProskomma's `blocks(...)` selection, so they're skipped here.
"""
import gzip
import json

from byte_array import ByteArray
import succinct as succ
import unsuccinctify as uns


def _populated_byte_array(b64):
    ba = ByteArray()
    ba.from_base64(b64)
    return ba


def load_pkf(pkf_bytes):
    """pkf_bytes -> {docset_id, selectors, docs: {docId: {headers, mainId, sequences}}}"""
    succinct_ob = json.loads(gzip.decompress(pkf_bytes))

    enums = {
        category: _populated_byte_array(succinct_ob["enums"][category])
        for category in ("ids", "wordLike", "notWordLike", "scopeBits", "graftTypes")
    }
    enum_idx = succ.enum_indexes(enums)

    docs = {}
    for doc_id, succinct_doc in succinct_ob["docs"].items():
        sequences = {}
        for seq_id, seq in succinct_doc["sequences"].items():
            blocks = []
            for succinct_block in seq["blocks"]:
                bs_ba = _populated_byte_array(succinct_block["bs"])
                bg_ba = _populated_byte_array(succinct_block["bg"])
                c_ba = _populated_byte_array(succinct_block["c"])
                blocks.append({
                    "bs": uns.unsuccinctify_block_scope(enums, enum_idx, bs_ba),
                    "bg": uns.unsuccinctify_grafts(enums, enum_idx, bg_ba),
                    "c": uns.unsuccinctify_items(enums, enum_idx, c_ba),
                })
            sequences[seq_id] = {"type": seq["type"], "blocks": blocks}
        docs[doc_id] = {
            "headers": succinct_doc["headers"],
            "mainId": succinct_doc["mainId"],
            "sequences": sequences,
        }

    return {
        "docset_id": succinct_ob["id"],
        "selectors": succinct_ob["metadata"]["selectors"],
        "docs": docs,
    }

File: tools/pkf-decode-rs/src/load.rs
//! Port of tools/pkf-decode-py/load.py.

use std::collections::HashMap;
use std::io::Read;

use flate2::read::GzDecoder;
use serde_json::Value;

use crate::byte_array::ByteArray;
use crate::succinct::{Enums, EnumIndexes};
use crate::unsuccinctify::{unsuccinctify_block_scope, unsuccinctify_grafts, unsuccinctify_items};
use crate::item::Item;

pub struct Block {
    pub bs: String,
    pub bg: Vec<(String, String)>,
    pub c: Vec<Item>,
}

pub struct Sequence {
    pub type_: String,
    pub blocks: Vec<Block>,
}

pub struct Doc {
    /// Preserves original JSON key order — matters for startDocument's
    /// header-dumping loop, which must match decode.mjs's output order
    /// (JS objects and Python dicts both preserve insertion order; a
    /// HashMap here would not).
    pub headers: Vec<(String, String)>,
    pub main_id: String,
    pub sequences: HashMap<String, Sequence>,
}

impl Doc {
    pub fn header(&self, key: &str) -> Option<&str> {
        self.headers.iter().find(|(k, _)| k == key).map(|(_, v)| v.as_str())
    }
}

pub struct Loaded {
    pub docset_id: String,
    pub selectors: HashMap<String, String>,
    pub docs: Vec<(String, Doc)>, // preserves insertion order, unlike a HashMap
}

fn as_str_map(v: &Value) -> HashMap<String, String> {
    v.as_object()
        .expect("expected object")
        .iter()
        .map(|(k, v)| (k.clone(), v.as_str().unwrap_or_default().to_string()))
        .collect()
}

fn as_str_pairs(v: &Value) -> Vec<(String, String)> {
    v.as_object()
        .expect("expected object")
        .iter()
        .map(|(k, v)| (k.clone(), v.as_str().unwrap_or_default().to_string()))
        .collect()
}

pub fn load_pkf(pkf_bytes: &[u8]) -> Loaded {
    let mut decoder = GzDecoder::new(pkf_bytes);
    let mut json_str = String::new();
    decoder.read_to_string(&mut json_str).expect("gzip decompress failed");
    let succinct_ob: Value = serde_json::from_str(&json_str).expect("invalid JSON in pkf");

    let enums_ob = &succinct_ob["enums"];
    let enums = Enums {
        ids: ByteArray::from_base64(enums_ob["ids"].as_str().unwrap()),
        word_like: ByteArray::from_base64(enums_ob["wordLike"].as_str().unwrap()),
        not_word_like: ByteArray::from_base64(enums_ob["notWordLike"].as_str().unwrap()),
        scope_bits: ByteArray::from_base64(enums_ob["scopeBits"].as_str().unwrap()),
        graft_types: ByteArray::from_base64(enums_ob["graftTypes"].as_str().unwrap()),
    };
    let idx = EnumIndexes::build(&enums);

    let mut docs = Vec::new();
    let docs_ob = succinct_ob["docs"].as_object().expect("docs must be an object");
    for (doc_id, succinct_doc) in docs_ob.iter() {
        let mut sequences = HashMap::new();
        let seqs_ob = succinct_doc["sequences"].as_object().expect("sequences must be object");
        for (seq_id, seq) in seqs_ob.iter() {
            let type_ = seq["type"].as_str().unwrap().to_string();
            let mut blocks = Vec::new();
            for succinct_block in seq["blocks"].as_array().expect("blocks must be array") {
                let bs_ba = ByteArray::from_base64(succinct_block["bs"].as_str().unwrap());
                let bg_ba = ByteArray::from_base64(succinct_block["bg"].as_str().unwrap());
                let c_ba = ByteArray::from_base64(succinct_block["c"].as_str().unwrap());
                blocks.push(Block {
                    bs: unsuccinctify_block_scope(&enums, &idx, &bs_ba),
                    bg: unsuccinctify_grafts(&enums, &idx, &bg_ba),
                    c: unsuccinctify_items(&enums, &idx, &c_ba),
                });
            }
            sequences.insert(seq_id.clone(), Sequence { type_, blocks });
        }
        docs.push((
            doc_id.clone(),
            Doc {
                headers: as_str_pairs(&succinct_doc["headers"]),
                main_id: succinct_doc["mainId"].as_str().unwrap().to_string(),
                sequences,
            },
        ));
    }

    Loaded {
        docset_id: succinct_ob["id"].as_str().unwrap().to_string(),
        selectors: as_str_map(&succinct_ob["metadata"]["selectors"]),
        docs,
    }
}
